### PR TITLE
Feature enable ingress for path based routing

### PR DESCRIPTION
**Pull request by kandrio**
**What this PR does / why we need it**:

This PR introduces path-based serving. It is related to the following issues:
- https://github.com/kserve/kserve/issues/824
- https://github.com/kserve/kserve/issues/2257

**Type of changes**
Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

**Feature/Issue validation/testing**:

I added tests in the source code for every change I introduced.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by yuzisun** [comment]
@kandrio Can you help rebase the code from master?

**Comment by dimara** [comment]
@yuzisun I will continue the work that @kandrio started. I will do the rebase ASAP and ping here once done! Thanks!

**Comment by dimara** [comment]
@yuzisun I rebased of the original PR and resolved the conflicts. Are there any other blockers?

**Comment by dimara** [comment]
@yuzisun any new on this? cc @adriangonz @Iamlovingit

**Comment by lizzzcai** [comment]
Hi @dimara, thanks for your follow-up work on this PR. I added my use case under the previous [issue](https://github.com/kserve/kserve/issues/2257#issuecomment-1223614792). It looks you have a separated `UrlTemplate` for the path-based url, have you consider reuse the existing `DomainTemplate`? (by working with `pathTemplate` to generate the Url).

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/url.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	"fmt"
+	"text/template"
+
+	"net/url"
+
+	"github.com/kserve/kserve/pkg/apis/serving/v1beta1"
+	"knative.dev/pkg/apis"
+)
+
+type UrlTemplateValues struct {
```
Comment (lizzzcai) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/url.go]: Is it ok to support values from `annotations` and `labels` as well, similar to [DomainTemplate](https://github.com/kserve/kserve/blob/39ec2d19de1d856a9527b800fda8fb380bc9441f/pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/domain.go#L30-L36).

**Comment by dimara** [comment]
Hi @lizzzcai! First of all thanks for the feedback.

> It looks you have a separated UrlTemplate for the path-based url

Well I rebased the initial PR without making any important changes. I think your comment came after the PR was submitted. My bad I guess for not taking it into account before doing the rebase. Sorry for that.

> have you consider reuse the existing DomainTemplate? (by working with pathTemplate to generate the Url).

Isn't ``DomainTemplate`` used only in *raw* deployment, i.e., without KNative? Note that with the current implementation, the host-based routing that KNative sets up (based on its own config-network and config-domain) still works and as such we can reach the underlying KNative Services directly. With `urlTemplate` KServe will configure an extra/alternative "route" without messing up the existing configuration.

So based on the above, do we need to change the initial implementation or keep it as is? The alternative would be to make KServe take into account all the `domainTemplate`, `urlScheme`, `ingressDomain`, `pathTemplate` settings regardless the deployment type, right?

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/url.go (hunk 1 above)
Comment (dimara) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/url.go]: ACK! Let's decide whether we go with urlTemplate and if we do then I will add a fixup on top. Thanks!

**Comment by dimara** [comment]
@lizzzcai @yuzisun any news on this? How can we move forward?

**Comment by lizzzcai** [comment]
Hi @dimara , sorry for the late reply. I don't use the raw deployment mode but from what I see the `DomainTemplate` uses the `ingressDomain` to define its domain. 

In stead of using `ingressDomain` directly in this PR, maybe the user can define a `DomainTemplate: {{ .IngressDomain }}`, which same as the domain in knative config-domain, and use it with the `pathTemplate` to generate the full path based url `{{.DomainTemplate}}/{{.pathTemplate}}`. 

In this case, we are able to generalize the implementation on both raw deployment mode and serverless mode, and support path-based routing in raw deployment mode in the future as well.

However, from what I see the current host is getting from [knative](https://github.com/kserve/kserve/blob/3c93df286a053ea4d913201a31d972aa9058330f/pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go#L314) directly. So setting value to `DomainTemplate` can be `optional` (set to knative servicehost if it is empty), as you only check if UrlTempalte (pathTemplate) [exist](https://github.com/kserve/kserve/blob/3c93df286a053ea4d913201a31d972aa9058330f/pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go#L323) to enable path-based routing.

what do you think?

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go [hunk 2]
```diff
@@ -311,6 +320,88 @@ func createIngress(isvc *v1beta1.InferenceService, config *v1beta1.IngressConfig
 		hosts = append(hosts, serviceHost)
 		gateways = append(gateways, config.IngressGateway)
 	}
+	if config.UrlTemplate != "" {
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go]: Can we add the test cases for the generated vs matching rules?

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go [hunk 3]
```diff
@@ -125,6 +125,15 @@ func getServiceUrl(isvc *v1beta1.InferenceService, urlScheme string, disableIsti
 	}
 }
 
+func contains(s []string, e string) bool {
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go]: We can move this to the utils file

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go [hunk 4]
```diff
@@ -311,6 +320,88 @@ func createIngress(isvc *v1beta1.InferenceService, config *v1beta1.IngressConfig
...
+				Headers: &istiov1alpha3.Headers{
+					Request: &istiov1alpha3.Headers_HeaderOperations{
+						Set: map[string]string{
+							"Host": network.GetServiceHostname(backend, isvc.Namespace),
+						},
+					},
+				},
+			})
+		} else if !contains(hosts, url.Host) {
+			// In this case, we have a host-based URL, so we add a host-based rule
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go]: I am confused about this case, in which scenario the `url.Host` is not in the `hosts` list?

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/url_test.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+				ingressConfig: &v1beta1.IngressConfig{
+					IngressDomain: "my.domain",
+				},
+			},
+			want: "",
+		},
+		{
+			name: "path based url template",
+			args: args{
+				name:      "model",
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/url_test.go]: is this the service name or model name? There could be multiple models in a given InferenceService

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go [hunk 6]
```diff
@@ -311,6 +320,88 @@ func createIngress(isvc *v1beta1.InferenceService, config *v1beta1.IngressConfig
...
+						Authority: &istiov1alpha3.StringMatch{
+							MatchType: &istiov1alpha3.StringMatch_Regex{
+								Regex: constants.HostRegExp(url.Host),
+							},
+						},
+						Gateways: []string{config.IngressGateway},
+					},
+				},
+				Rewrite: &istiov1alpha3.HTTPRewrite{
+					Uri: "/",
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go:362]: I am not sure how this is going to work, for example if the `UrlTemplate` is `"https://{{ .IngressDomain }}/{{ .Namespace }}/{{ .Name }}` and the user is calling with the following url `http://istio-ingressgateway.istio-system/user/isvc/v2/models/cifar/infer`, `url.path` would return `user/isvc` and rewrite to `/`.

Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go]: Does this only replace the matched prefix portion?

**Comment by yuzisun** [comment]
Thanks @kandrio for the great work and apologize for the late review.

**Comment by yuzisun** [comment]
@kandrio @lizzzcai Great discussions! My two cents is that we probably should not set `DomainTemplate` and `IngressDomain` in isvc configuration for serverless mode as they are only supposed to be used with `RawDeployment`, otherwise we may get into inconsistent state between Knative and KServe. In serverless mode we take the host directly from Knative so for path based routing we only need the `PathTemplate` part really, we can use this to decide whether to inject path based routing rules or not. On the other hand `RawDeployment` is implemented with Kubernetes ingress resource which does not support url rewrite, we might need to consider using [Gateway API](https://gateway-api.sigs.k8s.io/) in future to normalize the implementation between `Serverless` and `RawDeployment` mode.

**Comment by dimara** [comment]
> However, from what I see the current host is getting from knative directly.

Yes. So `domainTemplate` etc. are currently ignored in the default deployment type (serverless / with KNative). 

> So setting value to DomainTemplate can be optional.

We must think of existing deploymens here, since `domainTemplate` currently has a default value `"{{ .Name }}-{{ .Namespace }}.{{ .IngressDomain }}"`. And users may not have set this at all. So it is not clear when to use the KNative domain or use the one generated based on KServe configuration.

What we can do is have `pathTemplate` optional (empty by default, same as we do in this PR), and if this is set then it will use the `domainTemplate`, `ingressDomain, and `urlScheme`. So if a user wants to use path-based serving, which is not the default, they will have to specify something like:
```
        "ingressDomain"  : "mydomain.com",
        "domainTemplate": "{{ .IngressDomain }}",
        "urlScheme": "https",
        "pathTemplate": "/serving/{{ .Namespace }}/{{ .Name }}"
```
If `pathTemplate` is not set, then the serverless deployment type will continue to work as it does now, i.e., obtain the URL from Knative.

Another option would be to have a new separate boolean config, e.g., `generateURL` or something, that will be `false` by default. And only if the user sets this then all the above values will be taken into account. This way all existing options can be used for both raw and KNative deployments.

I think both options above are not that intrusive and do not need specific upgrade notes. @lizzzcai what do you think?

**Comment by lizzzcai** [comment]
Hi @dimara , thanks for your update.

> What we can do is have pathTemplate optional (empty by default, same as we do in this PR), and if this is set then it will use the domainTemplate, ingressDomain, and urlScheme`. 

from my understanding of the previous discussion,  `domainTemplate` will not be used in serverless mode even it is set.

I agreed with @yuzisun  idea. Just add `PathTemplate` should be enough and it will only work in serverless mode (need to document that it will not work for raw deployment mode), and inject it if the value is set. `PathTemplate` should support value from annotations and labels as well.

I agreed with the Gateway API idea as well, which can be implemented in the Raw deployment first to support path-based routing if there is any requirement.

**Comment by dimara** [comment]
>  In serverless mode we take the host directly from Knative so for path based routing we only need the PathTemplate part really, we can use this to decide whether to inject path based routing rules or not. 

@yuzisun @lizzzcai I believe we cannot use the host of KNative since this is specific to each isvc (includes both name and namespace). The idea is to keep KNative routing as is and have KServe to add support for path-based routing on top under a single domain. That is, have https://mydomain.com/serving/myns/myisvc point to KNative's internal http://myisvc.myns.svc.cluster.local. I understand configuring KNative's `domainTemplate` and `domain` is orthogonal. This will expose the underlying KNative Services.

For example, with current implementation, if we have KServe with:
```
urlTemplate: https://mydomain.com/serving/{{ .Namespace }}/{{ .Name }}
```
and KNative with default values for config-domain and domainTemplate, KServe will create a toplevel VirutalService like:
```
metadata:
  name: sklearn-iris
  namespace: kubeflow-user
spec:
  gateways:
  - knative-serving/knative-serving-cluster-gateway
  - knative-serving/knative-serving-gateway
  hosts:
  - sklearn-iris.kubeflow-user.svc.cluster.local
  - sklearn-iris-kubeflow-user.example.com
  - mydomain.com
  http:
  ...
   - headers:
      request:
        set:
          Host: sklearn-iris-predictor-default.kubeflow-user.svc.cluster.local
    match:
    - authority:
        regex: ^mydomain\.com(?::\d{1,5})?$
      gateways:
      - knative-serving/knative-serving-gateway
      uri:
        prefix: /serving/kubeflow-user/sklearn-iris/
    - authority:
        regex: ^mydomain\.com(?::\d{1,5})?$
      gateways:
      - knative-serving/knative-serving-gateway
      uri:
        exact: /serving/kubeflow-user/sklearn-iris
    rewrite:
      uri: /
    route:
    - destination:
        host: knative-serving-cluster-ingressgateway.knative-serving.svc.cluster.local
        port:
          number: 80
      weight: 100
```

This means that the user will be able to hit the ISVC with `https://mydomain.com/serving/kubeflow-user/sklearn-iris/v1/models/sklearn-iris:predict` or by port-forwarding knative-serving-ingressgateway and setting the `Host` header to `sklearn-iris-kubeflow-user.example.com`.


> My two cents is that we probably should not set DomainTemplate and IngressDomain in isvc configuration for serverless mode as they are only supposed to be used with RawDeployment
 
This is what argued before as well and that is why we went for a new option.
 
> otherwise we may get into inconsistent state between Knative and KServe.

Well with current implementation KNative routing continues to work as is. All we do is add an **extra** route on top to support path-based routing. As I said above, my understanding is that we cannot use the host of the underlying KNative service since this includes both name and namespace. @yuzisun @lizzzcai do I miss something here?

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go (hunk 2 above)
Comment (dimara) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go]: You mean extend `ingress_reconciler_test.go` to test createIngress() for path-based routing scenarios?

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go (hunk 4 above)
Comment (dimara) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go]: Well this is a corner case where we have set a host-based url template:
```
urlTemplate: https://{{ .Name }}-{{ .Namespace}}.serving.mydomain.com/
```
This is more or less the case where we configure KServe to expose the ISVC and let KNative with default values (example.com domain).

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/url_test.go (hunk 5 above)
Comment (dimara) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/url_test.go]: This is service (isvc) name. 
> There could be multiple models in a given InferenceService

ACK. I believe the value here, `model`, was taken from the existing `domain_test.go`. But if you want we can change it to something else.

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go (hunk 6 above)
Comment (dimara) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go]: Yes. For example:
```
/serving/myns/myisvc/v1/models/myisvc:predict
```
will become
```
/v1/models/myisvc:predict
```

Note that we end up having two rules; one with prefix one with exact match:
```
    - ...
      uri:
        prefix: /serving/kubeflow-user/sklearn-iris/
    - ...
      uri:
        exact: /serving/kubeflow-user/sklearn-iris
    rewrite:
      uri: /
```

**Comment by yuzisun** [comment]
@kandrio I think only configuring the domain in kserve and leaving the knative with default domain name will not work as top level of virtual service routes to the vs created by the knative, so the domain needs to be synced between knative and kserve in order for mydomain.com to work which is why we should configure the domain in knative instead of kserve for serverless mode.

**Comment by yuzisun** [comment]
I guess what you are saying is that we can route to the knative provisioned virtual service with in cluster domain so we do not actually need to use the external domain configuration in Knative right ?

**Comment by dimara** [comment]
> so the domain needs to be synced between knative and kserve in order for mydomain.com to work which is why we should configure the domain in knative instead of kserve for serverless mode.

The problem is that you cannot configure KNative to use a single domain (path-based routing). It uses host-based routing for all the VirutalServices it creates.

> route to the knative provisioned virtual service with in cluster domain so we do not actually need to use the external domain configuration in Knative right ?

Right. As described in https://github.com/kserve/kserve/pull/2357#issuecomment-1315129192, `https://mydomain/serving/kubeflow-user/sklearn-iris` will be routed to `knative-serving-cluster-ingressgateway.knative-serving.svc.cluster.local` using Host `sklearn-iris-predictor-default.kubeflow-user.svc.cluster.local`.
Still, you can configure KNative config-domain and domainTemplate, it if you want to have the underlying KSVC (predictor, transformer) be directly accessible. But these will always use host-based routing.

So based on the above, should we proceed with the existing implementation (`urlTemplate`) or use a `pathTemplate` as described in https://github.com/kserve/kserve/pull/2357#issuecomment-1313279838. @yuzisun @lizzzcai what do you think?

**Comment by lizzzcai** [comment]
> The problem is that you cannot configure KNative to use a single domain (path-based routing). It uses host-based routing for all the VirutalServices it creates.
Hi @dimara , if knative and kserve vs you created above in this PR using the same domain, for example:
```
# domain: "{{ .Name }}-{{ .Namespace }}.example.com"
sklearn-iris-default.example.com # knative
sklearn-iris-default.example.com/default/sk-learn-iris # kserve path-based routing, assume that sharing the same domain with knative.
```
In this case, if user send a request to `sklearn-iris-default.example.com/default/sk-learn-iris/v1/model/<name>:predict`, the request will be routed to knative endpoint rather than kserve path based endpoint based on the matching logic?

If this is the case, I will still prefer to use `pathTemplate` with the `domainTemplate` for path-based routing, future we can support in raw deployment as well. `urlTemplate` will make `domainTemplate` a bit redundant.

**Comment by dimara** [comment]
Let me try explain a bit more. Let's assume we create a `sklearn-iris` ISVC in the kubeflow-user namespace. KServe will create:
* `sklearn-iris-predictor-default` ksvc (KNative Service)
* `sklearn-iris` Istio VirtualService
* `sklearn-iris` Service

Until now, the KServe VirtualService was rewriting the request to reach the underlying KNative VirtualService. Specifically:
1. KServe looks at the status of the KNative service that reported a url like http://sklearn-iris-predictor-default-kubeflow-user.example.com
2. removes the `predictor-default` (see https://github.com/kserve/kserve/blob/faffcc2f4e026902012fbf00c28a0b16c5a9a29e/pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go#L87)
3. uses the result in the `hosts` of the VirtualService
4. sets the `Host` header to the **internal** host of the KNative Service, that is sklearn-iris-predictor-default.kubeflow-user.svc.cluster.local
5. re-routes the request to the KNative cluster ingressgateway.

This PR changes the Istio VirtualService to support path-based routing, i.e., use **per-isvc separate prefix** under a **single** domain. Using the KNative domain does not help since this is different per isvc. The path-based support comes on-top, that is host-based rules based on KNative Services are also added and host-based routing continues to work.

> I will still prefer to use pathTemplate with the domainTemplate for path-based routing, 

I am OK with that, but as we mentioned above, currently `domainTemplate` is not supposed to work with serverless mode (KNative) but only in raw deployments. If we go with your suggestion we have to go with one of the options described in  https://github.com/kserve/kserve/pull/2357#issuecomment-1313279838 and as a result all existing options will be used for both raw and KNative deployments.

@lizzzcai @yuzisun what do you think? How can we move forward?

**Comment by yuzisun** [comment]
> I am OK with that, but as we mentioned above, currently `domainTemplate` is not supposed to work with serverless mode (KNative) but only in raw deployments. If we go with your suggestion we have to go with one of the options described in [#2357 (comment)](https://github.com/kserve/kserve/pull/2357#issuecomment-1313279838) and as a result all existing options will be used for both raw and KNative deployments.
> 
> @lizzzcai @yuzisun what do you think? How can we move forward?

I think `domain` and `domainTemplate` can still be configured in Knative and if `pathTemplate` exists then we can extract the domain from the predictor host to construct the url. In this way we can keep consistent that in serverless mode `domain` and `domainTemplate` is configured with Knative and in raw deployment it is configured with KServe.

**Comment by dimara** [comment]
> I think domain and domainTemplate can still be configured in Knative 

@yuzisun KNative uses host-based routing. So the domainTemplate will be in the format of `{{ .Name }}-{{ .Namespace }}.example.com` which mean a **different domain per-isvc**. So we cannot have a single domain and use path-based routing as this PR aims for.

If we have a `domainTemplate: example.com` in KNative nothing will work since KNative will end up creating VirtualServices with **conflicting** hosts. For example if I create two isvc, `sklearn-iris` and `sklearn-iris-2`, I will get two KNative services that will create two VirtualServices:
* sklearn-iris-predictor-default-ingress
* sklearn-iris-2-predictor-default-ingress

The first will have:
```
  hosts:
  - example.com
  - sklearn-iris-predictor-default.kubeflow-user
  - sklearn-iris-predictor-default.kubeflow-user.svc.
  - sklearn-iris-predictor-default.kubeflow-user.svc.cluster.local
```
The second will have:
```
  - example.com
  - sklearn-iris-2-predictor-default.kubeflow-user
  - sklearn-iris-2-predictor-default.kubeflow-user.svc.
  - sklearn-iris-2-predictor-default.kubeflow-user.svc.cluster.local
```
And both will have a rule that will match:
```
  - match:
    - authority:
        prefix: example.com
      gateways:
      - knative-serving/knative-serving-gateway
```
but with different route:
```
    - destination:
        host: sklearn-iris-predictor-default-00002.kubeflow-user.svc.cluster.local
        port:
          number: 80
```

All KNative services will report the same URL and if you hit them you will not know where Istio will route you to. I tried that and actually the KSVC do not get ready:
```
$ kubectl get ksvc -A
NAMESPACE       NAME                               URL                  LATESTCREATED                            LATESTREADY                              READY     REASON
kubeflow-user   sklearn-iris-2-predictor-default   http://example.com   sklearn-iris-2-predictor-default-00001   sklearn-iris-2-predictor-default-00001   True      
kubeflow-user   sklearn-iris-predictor-default     http://example.com   sklearn-iris-predictor-default-00001     sklearn-iris-predictor-default-00001     Unknown   Uninitialized
```
And the `net-istio-controller` was complaining:
```
{"severity":"ERROR","timestamp":"2022-11-25T03:01:19.082636575Z","logger":"net-istio-controller.istio-ingress-controller","caller":"status/status.go:403","message":"Probing of http://example.com:80 failed, IP: 172.17.0.38:8080, ready: false, error: unexpected hash: want \"3008c6929ee505f2969ae52d7dee73e9b87aba272e3f48db84c107f43ec3c868\", got \"2f018fb32ed5c854768fd0c6bc5d00fdbaf75e3d665c030567025ce92de18863\
```
These hashes are the K-Network-Hash added by Istio VS. For my KSVC I got:
* sklearn-iris-predictor-default: 3008c6929ee505f2969ae52d7dee73e9b87aba272e3f48db84c107f43ec3c868
* sklearn-iris-2-predictor-default: 2f018fb32ed5c854768fd0c6bc5d00fdbaf75e3d665c030567025ce92de18863

In other words, KNative tries to probe sklearn-iris-predictor-default and got as a response the hash ofr sklearn-iris-2-predictor-default. Once I restored `domainTemplate: '{{.Name}}-{{.Namespace}}.{{.Domain}}'` both KSVC became ready.

So if we want path-based routing under a **single** domain, we have to do that in KServe on top of KNative's host-based routing. This is what this PR does. Am I missing something here?

**Comment by yuzisun** [comment]
@kandrio I do not think we need to modify the `domainTemplate` in Knative, we can extract the domain from the predictor host in inference service status. For example KServe can get the domain `example.com` from `sklearn-iris-user1.default.example.com` using the `domainTemplate` as the parser, all we need to know is the domain. In knative the domain is configured [here ](https://github.com/knative/serving/blob/main/config/core/configmaps/domain.yaml), alternatively we can get this directly by reading the knative domain configmap.

**Comment by dimara** [comment]
@yuzisun first of all sorry for the late reply. Let me go through your comments/suggestions.

> For example KServe can get the domain example.com from sklearn-iris-user1.default.example.com using the domainTemplate as the parser, all we need to know is the domain. 

Oh. Now I get what you meant. I got confused when you said:

>  we can extract the domain from the predictor host to construct the url.

I thought with "extract" you meant take the predictor host as we currently do.

So to summarize your suggestion is to:
1. Introduce a `pathTemplate` config in inferenceservice-config of KServe.
2. This will be used in serverless mode for now.
3. If `pathTemplate` is set, the controller will obtain/compute the domain from KNative and use it to configure the toplevel VirtualService

My objection regarding (3) is that how Knative Services are exposed is kinda orthogonal to how KServe exposes their Inference service via the **top-level** VirtualService. KNative may use a completely different domain (or keep the dummy example.com) or choose not to expose its Services at all (if for example having a wildcard certificate and DNS record for the KNative URLs may not be an option). Plus, making KServe depend 100% on KNative would not allow us to have the two modes (raw, serverless) share the same expose logic in the future.

Leaving my objection aside, I believe there are some implementation issues if we try to use the domain from KNative. Specifically, you suggested to have the controller inspect the KNative configuration:
> we can get this directly by reading the knative domain configmap.

This is bit problematic since you will have to allow KServe controller to read confimaps in other namespaces. Plus, KNative may decide to change the structure of these configmaps and this will break KServe. Apart from that, this configmap is kinda complicated since it allows for specifying different domains based on selectors so parsing it may not be that trivial.

You also suggested:
>  get the domain from the host using the domainTemplate as the parser

While reading domainTemplate from config-network is easier (but with the same RBAC and compatibility issues mentioned above), To use the Go Template in a reverse manner (see 
https://github.com/golang/go/issues/52473), you have to do some tricks like rendering the template without the domain and compare the result with the status.url of the KNative service.


That said, I think we should decouple from how KNative exposes their Services. This PR supports exposing the InferenceService in a manner that the underlying KNative services don't need to be exposed at all. So I would either go with the initial implementation (with urlTemplate) or with what described in https://github.com/kserve/kserve/pull/2357#issuecomment-1313279838 (with pathTemplate). @yuzisun @lizzzcai What do you think?

**Comment by dimara** [comment]
Note that we already have support for setting the `urlScheme` in serverless deployments (see https://github.com/kserve/kserve/pull/2202). This is in-line with our suggestion where we have KServe decide on the final InferenceService URL, regardless the deployment mode. Specifically, we can use the `ingressDomain` along with `pathTemplate` and `urlScheme` to construct the final URL (as suggested in https://github.com/kserve/kserve/pull/2357#issuecomment-1313279838).

**Comment by yuzisun** [comment]
> My objection regarding (3) is that how Knative Services are exposed is kinda orthogonal to how KServe exposes their Inference service via the **top-level** VirtualService. KNative may use a completely different domain (or keep the dummy example.com) or choose not to expose its Services at all (if for example having a wildcard certificate and DNS record for the KNative URLs may not be an option). Plus, making KServe depend 100% on KNative would not allow us to have the two modes (raw, serverless) share the same expose logic in the future.

I am ok with configuring `ingressDomain` in KServe for path based routing, but I think we need to carefully document the configurations as currently in serverless mode the domain has to be configured in Knative when you use host based urls.

**Comment by yuzisun** [comment]
@kandrio thanks for the awesome work! to wrap up the PR I think we still need to add an e2e test for the path based routing.

**Comment by dimara** [comment]
@yuzisun just found time to finally switch to this. I will ping here once I have the PR ready with all the changes we discussed above. 
> we still need to add an e2e test for the path based routing.

OK. I'll try to add one.

**Comment by dimara** [comment]
@yuzisun I am done with the refactor. Here is a summary of what I ended up implementing:

* introduce `pathTemplate` in `ingressConfig`
* if `pathTemplate` is given require `ingressDomain`
* Add `path.go` (similar to `domain.go`) that will be responsible for rendering / validating `pathTemplate` given an isvc. Add unittests for this.
* Have `getServiceUrl()` take into account `pathTemplate`. If this is set then the `isvc.status.url` will end up having a path-based URL.
* Extend `test-fast` e2e test, to also test path-based routing.
* if `pathTemplate` is set then the virtualservice will obtain a new rule for the path (as it did in the initial PR) and a host for `ingressDomain`.
* the new version seems to be simpler and more straightforward than then initial implementation since we don't have to cover all corner cases with a user provided urlTemplate, e.g., one that may be host-based.

I noticed that the `test-fast` and `test-helm` of the e2e tests failed: https://github.com/kserve/kserve/actions/runs/3947181022/

But what failed seems totally irrelevant. For example, the minikube-setup step failed with:
```
Error from server (InternalError): Internal error occurred: failed calling webhook "config.webhook.serving.knative.dev": failed to call webhook: Post "https://webhook.knative-serving.svc:443/config-validation?timeout=10s": dial tcp 10.108.221.58:443: connect: connection refused
783
```
I understand that `kubectl patch cm config-deployment -n knative-serving` failed because the webhook refused the connection even though we wait for this to become ready.  This seems to be a race. For example the same step succeeded in `test-slow` job. Maybe related with istio sidecar running in this node. Can we re-run this?

Any idea why the test-helm job failed? It seems port-forward to the predictor pod failed:
```
E0118 10:02:59.708333   11169 portforward.go:400] an error occurred forwarding 8008 -> 8008: error forwarding port 8008 to pod 3b7d1ad3ce8eccdfc8e66f1db4599ba5a9ded6c9b4527f7393ee8b08013743be, uid : exit status 1: 2023/01/18 10:02:59 socat[13582] E connect(5, AF=2 127.0.0.1:8008, 16): Connection refused
```
This seems to be irrelevant with this PR, right? Since it hits the endpoint directly from inside the predictor. Can we ignore this?

@yuzisun Assuming the e2e tests pass, is there any other blocker for merging this PR? cc @lizzzcai

**Comment by yuzisun** [comment]
Thanks @kandrio for the update! I will go through the PR again, we are making a few fixes for the flaky tests, meanwhile I have rerun the tests for your PR.

**Comment by dimara** [comment]
> Thanks @kandrio for the update!

Just for the record I am @dimara :) @kandrio was the one that initially submitted the PR but he had to drop so we continued his work.

> we are making a few fixes for the flaky tests

ACK.
> meanwhile I have rerun the tests for your PR.

OK. Let me know if you have any comments about the revised PR, then. Thanks!

**Comment by dimara** [comment]
@yuzisun did you have the time to take a look?  I see that you re-submitted the tests and everything (including the test-fast that this PR extends) worked. Let me know if there are any blockers for merging this. Thanks!

File: pkg/apis/serving/v1beta1/configmap.go [hunk 7]
```diff
@@ -114,6 +116,19 @@ func NewIngressConfig(cli client.Client) (*IngressConfig, error) {
...
+			// * both Name and Namespace are used to avoid collisions
+			// * starts with a /
+			// For now simply check that this is a valid template.
+			_, err := template.New("path-template").Parse(ingressConfig.PathTemplate)
+			if err != nil {
+				return nil, fmt.Errorf("Invalid ingress config, unable to parse pathTemplate: %v", err)
+			}
+			if ingressConfig.IngressDomain == "" {
+				return nil, fmt.Errorf("Invalid ingress config - igressDomain is required if pathTemplate is given")
+			}
```
Comment (yuzisun) [inline pkg/apis/serving/v1beta1/configmap.go:130]: The `inferenceservice-config`  has the default value `example.com` so this check does not take effect , let's add a comment in the configmap that `IngressDomain` is used in case of `RawDeployment ` and path based routing for `Serverless` mode. https://github.com/kserve/kserve/blob/b15cd9e43771aa90ae8f2e0094c4333197af6bd1/config/configmap/inferenceservice.yaml#L62

**Comment by yuzisun** [comment]
@dimara Thanks for the awesome work to implement this long wanted feature!! Sorry for slow reviews.
I have created doc follow up issues
- https://github.com/kserve/kserve/issues/2678
- https://github.com/kserve/website/issues/225

/lgtm
/approve

File: pkg/apis/serving/v1beta1/configmap.go (hunk 7 above)
Comment (dimara) [reply pkg/apis/serving/v1beta1/configmap.go:130]: @yuzisun isn't the default value assigned a few lines below if this is not given?

https://github.com/kserve/kserve/blob/c587ef1769f33f6134ba16b6ef1e7f569eb3eda4/pkg/apis/serving/v1beta1/configmap.go#L138

**Comment by zamonia500** [comment]
Does this PR cover gRPC inference too?
